File: framework/languages/dotnet/samples/Bingo/Server/Play/Infrastructure/ZLink/Actors/PlayerActorTransferAdapter.cs
using Zlink.Framework.Contracts.Actors;
using Zlink.Framework.Contracts.Messaging;
using Zlink.Framework.Contracts.Spots;

namespace Bingo.Server.Play.Infrastructure.ZLink.Actors;

internal sealed class PlayerActorTransferAdapter : IZLinkActorTransferAdapter<PlayerActor>
{
    public ValueTask<ZLinkMessage> TransferOutAsync(
        PlayerActor actor,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        return ValueTask.FromResult(ZLinkMessage.From(new PlayerActorTransferState(
            actor.DisplayName,
            actor.RoomId,
            actor.DestroyAfterEntrySpotJoin,
            actor.Disconnected)));
    }

    public ValueTask<PlayerActor> TransferInAsync(
        string actorId,
        IZLinkActorContext context,
        ZLinkMessage state,
        CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();
        var transferred = state.Decode<PlayerActorTransferState>();
        var actor = new PlayerActor(actorId, context);
        actor.SetDisplayName(transferred.DisplayName);
        if (!string.IsNullOrEmpty(transferred.RoomId)) actor.JoinRoom(transferred.RoomId);
        if (transferred.DestroyAfterEntrySpotJoin) actor.MarkForDestroyAfterRoomLeave();
        if (transferred.Disconnected) actor.MarkDisconnected();
        return ValueTask.FromResult(actor);
    }

    private sealed record PlayerActorTransferState(
        string DisplayName,
        string RoomId,
        bool DestroyAfterEntrySpotJoin,
        bool Disconnected);
}

File: framework/languages/node/samples/Bingo.Ts/Server/Play/Infrastructure/ZLink/Actors/player-actor-transfer-adapter.ts
import { ZLinkMessage } from '@zlink-systems/framework';
import type { ZLinkActorTransferAdapter } from '@zlink-systems/framework';
import { PlayerActor } from './player-actor';
import { PlayerActorTransferState } from '../../../../../Shared/Contracts/bingo-messages.generated';

class PlayerActorTransferAdapter implements ZLinkActorTransferAdapter<PlayerActor> {
  async transferOut(actor: PlayerActor): Promise<ZLinkMessage> {
    return ZLinkMessage.from(new PlayerActorTransferState({
      displayName: actor.displayName,
      destroyAfterEntrySpotJoin: actor.destroyAfterEntrySpotJoin,
      disconnected: actor.disconnected
    }));
  }

  async transferIn(actorId: string, state: ZLinkMessage): Promise<PlayerActor> {
    const restored = state.decode<PlayerActorTransferState>();
    return new PlayerActor(
      actorId,
      restored.displayName,
      restored.destroyAfterEntrySpotJoin,
      restored.disconnected
    );
  }
}

export { PlayerActorTransferAdapter };
